### PR TITLE
Handle 429 errors from watch requests in reflector library.

### PR DESCRIPTION
**Pull request by wojtek-t**
/kind bug
/priority important-soon

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/client-go/tools/cache/reflector.go [hunk 1]
```diff
@@ -436,6 +437,10 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
 					klog.Warningf("%s: watch of %v ended with: %v", r.name, r.expectedTypeName, err)
 				}
 			}
+			if apierrors.IsTooManyRequests(err) {
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/tools/cache/reflector.go:440]: would it make sense to put this in the switch statement and log differently?

File: staging/src/k8s.io/client-go/tools/cache/reflector_test.go [hunk 2]
```diff
@@ -436,6 +436,59 @@ func TestReflectorListAndWatchInitConnBackoff(t *testing.T) {
...
+	clock clock.Clock
+	calls int
+}
+
+func (f *fakeBackoff) Backoff() clock.Timer {
+	f.calls++
+	return f.clock.NewTimer(time.Duration(0))
+}
+
+func TestBackoffOnTooManyRequests(t *testing.T) {
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/tools/cache/reflector_test.go:449]: does this exercise both code paths introduced above?

Comment (wojtek-t) [reply staging/src/k8s.io/client-go/tools/cache/reflector_test.go:449]: Yes - the case 0 exercises the path of error from the watch requests, case 1 (in the switch in watch func) exercises the "watch event of type error".

File: staging/src/k8s.io/client-go/tools/cache/reflector.go (hunk 1 above)
Comment (wojtek-t) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go:440]: good point - done

**Review by wojtek-t — COMMENTED** [review]
@liggitt - thanks for comments, PTAL

File: staging/src/k8s.io/client-go/tools/cache/reflector.go [hunk 3]
```diff
@@ -432,6 +433,10 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
 					// has a semantic that it returns data at least as fresh as provided RV.
 					// So first try to LIST with setting RV to resource version of last observed object.
 					klog.V(4).Infof("%s: watch of %v closed with: %v", r.name, r.expectedTypeName, err)
+				case apierrors.IsTooManyRequests(err):
+					klog.V(4).Infof("%s: watch of %v returned 429 - backing off", r.name, r.expectedTypeName)
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/tools/cache/reflector.go:437]: not sure about verbosity here... this seems like it should be more visible than an expired error

Comment (wojtek-t) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go]: That makes sense - it shouldn't be that frequent. Switched to V(2)

**Comment by wojtek-t** [comment]
PTAL

**Comment by jan--f** [comment]
Hi @wojtek-t it would be great if this could be backported to 1.21.I'd be more than happy to help with that, wdyt?

**Comment by wojtek-t** [comment]
> Hi @wojtek-t it would be great if this could be backported to 1.21.I'd be more than happy to help with that, wdyt?

This is in a grey area to me - it's not a typical correctness bug - it's rather performance inefficiency.
We usually don't backport those kind of changes (unless they are regressions, which isn't the case here).

Wearing my scalability hat, I would like to backport it, but wearing my approver hat, I don't think we have strong enough reason for it.
@liggitt - WDYT ^^

**Comment by liggitt** [comment]
I wouldn't recommend we backport this

**Comment by wojtek-t** [comment]
> I certainly don't fully understand the code related to informers but I also didn't see how retries happen on 429 errors and this PR seemed to fit the bill.
Is this fix really just a performance improvement?

Without this PR - we fail ListAndWatch - but that is called in a loop as part of Run:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/reflector.go#L220

So yes - it's pretty much performance inefficiency catch - try to re-watch, instead of forcing relist.

**Pull request merged**